[16.0][ADD] hr_timesheet_overtime_rate: Add hours_worked field

carmenbianca opened the pull request:
## Description

This should hopefully fix the problem described in https://github.com/coopiteasy/cie-timesheet/pull/33#issuecomment-2209048295

I'm not 100% about this PR at time of writing:

- The summary page doesn't work.
- ~~I feel like everything related to different rates on certain days is not at all related to the rest of this module. This is a part of that. Maybe it should be factored out?~~ Done.

## Odoo task (if applicable)

T12629

## Checklist before approval

- [ ] Tests are present (or not needed).
- [ ] Credits/copyright have been changed correctly.
- [ ] Change log snippet is present.
- [ ] (If a new module) Moving this to OCA has been considered.

carmenbianca committed:
[REF] hr_timesheet_overtime: Add hours_worked field

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[ADD] hr_timesheet_overtime_rate: Factored out of hr_timesheet_overtime

This functionality was completely unlinked from the rest of
hr_timesheet_overtime.

I would have done this in two commits, but then the files wouldn't move
over cleanly, losing history.

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[REM] hr_timesheet_overtime_rate: Make summary view read-only

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[FIX] hr_timesheet_overtime_rate: Populate hours_worked on installation

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca commented:
@huguesdk ready for review

carmenbianca committed:
[REF] hr_timesheet_overtime_rate: Rate acquisition is no longer an @api.model method

This makes it easier to extend with additional available information on
the record.

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

huguesdk commented on hr_timesheet_overtime_rate/__manifest__.py:
```diff
@@ -0,0 +1,29 @@
+{
+    "name": "Overtime Rate",
+    "summary": """
+        Define an extra rate for working certain days.""",
```
```suggestion
    "summary": "Define an extra rate for working certain days",
```

huguesdk commented on hr_timesheet_overtime_rate/__manifest__.py:
```diff
@@ -0,0 +1,29 @@
+        "views/hr_timesheet_sheet_views.xml",
+    ],
+    "demo": [],
+    "qweb": [],
```
please remove default values.

```suggestion
    "depends": [
        "hr_timesheet_sheet",
    ],
    "data": [
        "security/ir.model.access.csv",
        "data/data.xml",
        "views/resource_views.xml",
        "views/account_analytic_line_views.xml",
        "views/hr_timesheet_sheet_views.xml",
    ],
```

huguesdk commented on hr_timesheet_overtime_rate/models/account_analytic_line.py:
```diff
@@ -0,0 +1,71 @@
+    def get_overtime_rate(self):
+        self.ensure_one()
+        # n.b. from_string also works on date objects, returning itself.
+        weekday = fields.Date.from_string(self.date).weekday()
```
why not then simply:
```suggestion
        weekday = self.date.weekday()
```
?

huguesdk commented on hr_timesheet_overtime_rate/readme/ROADMAP.rst:
```diff
@@ -0,0 +1,17 @@
+robust.
+
+Because ``unit_amount`` is computed from ``hours_worked``, this module is not
+compatible with ``hr_timesheet_begin_end``.To make this module subsequently
```
```suggestion
compatible with ``hr_timesheet_begin_end``. To make this module subsequently
```

huguesdk commented on hr_timesheet_overtime_rate/__manifest__.py:
since this module doesn’t depend on `hr_timesheet_overtime`, shouldn’t it be named something like `hr_timesheet_worktime_rate`? all its ui mentions “overtime rate”, though.

carmenbianca replied:
You're technically right, but this would require more refactoring and migration scripts. I propose we don't do this because there is no real need for this at the moment.

huguesdk commented on hr_timesheet_overtime_rate/__manifest__.py:
```diff
@@ -0,0 +1,29 @@
+    "license": "AGPL-3",
+    "application": False,
+    "depends": [
+        "hr_timesheet_sheet",
```
shouldn’t this be `hr_timesheet`?

edit: ok, no, because it overrides views for `hr_timesheet_sheet` too.

huguesdk reviewed (changes requested):
great work! almost there.

note that the `hours_worked` field doesn’t appear in the timesheet lines of tasks (on the task form). i think that this should be added.

carmenbianca committed:
[IMP] hr_timesheet_overtime_rate: Improve code aesthetics

No functional differences; code is a little prettier.

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[IMP] hr_timesheet_overtime_rate: Decrease priority of views

By decreasing the priority, we make sure that hours_worked is as close
to unit_amount as possible.

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[IMP] hr_timesheet_overtime_rate: Add hours_worked to task view

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

huguesdk reviewed (approved):
yay, lgtm! 🎉
